tests/framework/static_analysis.py

```diff
@@ -0,0 +1,597 @@
+        affected_registers = [
+            match
+            for (_, match, _) in re.findall(r"(\s|,)([wx]\d{1,2})(\W|)", str(insn))
+        ]
+        if reg not in affected_registers:
+            return reg
+
+        match insn.mnemonic:
+            case "mov":
+                if len(insn.args) != 2:
+                    raise UnsupportedInstructionError(insn, reg)
+
+                dst, src = insn.args
+
+                if dst == reg:
+                    # an immediate load
+                    if src.startswith("#"):
+                        return int(src[3:], 16)
+
+                    if src in ["xzr", "wzr"]:
+                        # See https://developer.arm.com/documentation/102374/0102/Registers-in-AArch64---other-registers
+                        return 0
+
+                    # We moved something into our target register. If it's a new register, we understand
+                    # what's going on. Anything else, and tough luck
+                    if re.match(r"^[xw]\d{1,2}$", src):
+                        return src
+
+                    raise UnsupportedInstructionError(insn, reg)
+                return reg
+            case "movk":
+                # https://developer.arm.com/documentation/dui0802/a/A64-General-Instructions/MOVK
+                assert len(insn.args) in [2, 3], str(insn)
+
+                immediate = int(insn.args[1][3:], 16)
+                shift = 0
+                if len(insn.args) == 3:
+                    # shift has form "lsl #<shift>", so strip first 5 characters
+                    shift = int(insn.args[2][5:])
+
+                mask = 0b1111_1111_1111_1111 << shift
+
+                return reg, lambda x: (x & ~mask) | (immediate << shift)
+            case "add" | "sub":
+                if len(insn.args) != 3:
+                    raise UnsupportedInstructionError(insn, reg)
+
+                dst, src, imm = insn.args
+
+                if dst != reg:
+                    return reg
+
+                try:
+                    # We can only handle additions of constants, because
+                    # the backpropagation algorithm cannot follow multiple registers.
+                    imm = int(imm[3:], 16)
+                except ValueError as exc:
+                    raise UnsupportedInstructionError(insn, reg) from exc
+
+                if insn.mnemonic == "add":
+                    return src, lambda x: x + imm
+                return src, lambda x: x - imm
+
+        raise UnsupportedInstructionError(insn, reg)
+
+    def instruction_from_str(self, insn_str: str) -> "Instruction":
+        parts = insn_str.split(maxsplit=1)
+
+        args = []
+        if len(parts) > 1:
+            args = parts[1].split(",")
+
+            # Strip each argument, in case objdump decides to put spaces after commas
+            args = list(map(str.strip, args))
+            # The last argument might have a comment after it. Strip that off
+            args[-1] = args[-1].split("//")[0]
```

Comment:
Is the comment string different in x86 and aarch64?

Reply:
yeah :( On x86_64, the comment string is `#`. But on ARM, objdump uses a `#` prefix to indicate immediates, so the comment string is `//` instead :joy: (x86_64 uses `$` as the prefix for immediates)